src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -538,57 +539,67 @@ public void Reset()
-            // minus 5 for the max chunked prefix size and minus 2 for the chunked ending, leaving a total of
-            // 4089.
+            var originalSizeHint = sizeHint;
+
+            // To guarantee that sizeHint is a minimumSize, we need to call GetMemory with a sizeHint
+            // larger than what was provided. We add the begin and end chunk lengths here.
+            if (sizeHint > 0)
+            {
+                sizeHint += ChunkWriter.GetBeginChunkByteCount(sizeHint) + EndChunkLength;
+            }
 
             if (!_currentChunkMemoryUpdated)
             {
-                // First time calling GetMemory
-                _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
-                _currentChunkMemoryUpdated = true;
+                UpdateCurrentChunkMemory(sizeHint);
             }
 
-            var memoryMaxLength = _currentChunkMemory.Length - BeginChunkLengthMax - EndChunkLength;
-            if (_advancedBytesForChunk >= memoryMaxLength - sizeHint && _advancedBytesForChunk > 0)
+            if (_advancedBytesForChunk >= _currentChunkMemory.Length - _currentMemoryPrefixBytes - EndChunkLength - originalSizeHint && _advancedBytesForChunk > 0)
             {
-                // Chunk is completely written, commit it to the pipe so GetMemory will return a new chunk of memory.
                 var writer = new BufferWriter<PipeWriter>(_pipeWriter);
                 WriteCurrentMemoryToPipeWriter(ref writer);
                 writer.Commit();
-
                 _unflushedBytes += writer.BytesCommitted;
-                _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
-                _currentChunkMemoryUpdated = true;
+
+                UpdateCurrentChunkMemory(sizeHint);
             }
 
             var actualMemory = _currentChunkMemory.Slice(
-                BeginChunkLengthMax + _advancedBytesForChunk,
-                memoryMaxLength - _advancedBytesForChunk);
-
-            Debug.Assert(actualMemory.Length <= 4089);
+                _currentMemoryPrefixBytes + _advancedBytesForChunk,
+                _currentChunkMemory.Length - _currentMemoryPrefixBytes - EndChunkLength - _advancedBytesForChunk);
 
             return actualMemory;
         }
 
+        private void UpdateCurrentChunkMemory(int sizeHint)
+        {
+            _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
+            _currentMemoryPrefixBytes = ChunkWriter.GetBeginChunkByteCount(_currentChunkMemory.Length);
+
+            // Super edge case. If we call GetMemory and it returns 4099, we would originally calculate _currentMemoryPrefixBytes
+            // as 6 bytes. However, after subtracting the bytes for _currentChunkMemory.Length and the endPrefix, the real
+            // body length would be 5 bytes. 
+            _currentMemoryPrefixBytes = ChunkWriter.GetBeginChunkByteCount(_currentChunkMemory.Length - _currentMemoryPrefixBytes - EndChunkLength);
```

Comment:
Couldn't this logic be internal to GetBeginChunkByteCount?

Reply:
Do you even whether this logic is stable? If the next line was also

```C#
_currentMemoryPrefixBytes = ChunkWriter.GetBeginChunkByteCount(_currentChunkMemory.Length - _currentMemoryPrefixBytes - EndChunkLength);
```

Is it guaranteed that _currentMemoryPrefixBytes won't change again?